openedx/core/lib/content_security_policy/middleware.py

```diff
@@ -0,0 +1,149 @@
+            reporting_suffix += f"; report-to {reporting_endpoint_name}"
+
+    def clean_header(value):
+        # Collapse any internal whitespace that contains a newline. This allows
+        # writing the setting value as a multi-line string, which is useful for
+        # CSP -- the values can be quite long.
+        value = re.sub("\\s*\n\\s*", " ", value).strip()
+        # Remove any trailing semicolon, which we allow (for convenience).
+        # The CSP spec does not allow trailing semicolons or empty directives.
+        value = re.sub("[;\\s]+$", "", value)
+        return value
+
+    if enforce_policies:
+        headers['Content-Security-Policy'] = clean_header(enforce_policies) + reporting_suffix
+
+    if report_policies:
+        headers['Content-Security-Policy-Report-Only'] = clean_header(report_policies) + reporting_suffix
+
+    return headers
+
+
+def _append_headers(response_headers, more_headers):
+    """
+    Append to the response headers. If a header already exists, assume it is
+    permitted to be multi-valued (comma-separated), and update the existing value.
+
+    Arguments:
+        response_headers: response.headers (or any dict-like object), to be modified
+        more_headers: Dict of header names to values
+    """
+    for k, v in more_headers.items():
+        if existing := response_headers.get(k):
+            response_headers[k] = f"{existing}, {v}"
+        else:
+            response_headers[k] = v
+
+
+class MiddlewareNotUsed(Exception):
+    """This middleware is not used in this server configuration"""
+
+
+def _get_override(request, url_specific_csps):
+    """
+    Check the CUSTOM_CSPS environment variable for a match to the request url.
+    If a regex is included that matches the request url, that means that the CSP header
+    should be overwritten with the provided string for this url.
+    """
+    if not url_specific_csps or not isinstance(url_specific_csps, list):
+        return None
+    # TODO: handle errors for when env variable has wrong format
+    for csp in url_specific_csps:
+        if not isinstance(csp, list) or len(csp) != 2:
+            continue
+        regex, custom_csp = csp
+        if re.search(regex, request.path):
+            return custom_csp
+    return None
+
+
+def content_security_policy_middleware(get_response):
+    """
+    Return middleware handler based on CSP headers.
+    These are specified in the environment variables `CSP_STATIC_ENFORCE`,
+    `CSP_STATIC_REPORT_ONLY`, `CSP_STATIC_REPORTING_URI`, and `CSP_STATIC_REPORTING_NAME`.
+
+    It is possible to override the CSP headers for specific URLs by setting the
+    `CUSTOM_CSPS` environment variable to a list of tuples. Each tuple should
+    contain a regex and a string. If the regex matches the request URL, the
+    CSP header part specified in `CSP_STATIC_ENFORCE` will be overwritten with the provided string.
```

Comment:
This comment should be replaced with something like:
```
GET_CUSTOM_CSPS is defined in lms/common.py and cms/common.py and returns custom CSP for the specific service.
```

Reply:
updated